File: native-bridge/android/PuttTelemetry.kt
package com.puttvision.bridge

import org.json.JSONObject

/**
 * Versioned shot-result contract shared with the Unity renderer.
 *
 * Camera/vision code stays native. Publish one instance only after native validation has
 * produced a final shot result; do not use this object for raw frame transport.
 */
data class PuttTelemetry(
    val schemaVersion: Int = CURRENT_SCHEMA_VERSION,
    val shotId: String,
    val timestampNs: Long,
    val ballSpeedMps: Float,
    val launchDirectionDeg: Float,
    val faceAngleDeg: Float = 0f,
    val pathAngleDeg: Float = 0f,
    val impactOffsetMm: Float = 0f,
    val confidence: Float = 0f,
    val validityFlags: Int = 0,
) {
    fun isSimulationReady(): Boolean =
        schemaVersion == CURRENT_SCHEMA_VERSION &&
            ballSpeedMps.isFinite() &&
            ballSpeedMps >= 0f &&
            launchDirectionDeg.isFinite()

    fun toJson(): String = JSONObject().apply {
        put("schemaVersion", schemaVersion)
        put("shotId", shotId)
        put("timestampNs", timestampNs)
        put("ballSpeedMps", ballSpeedMps.toDouble())
        put("launchDirectionDeg", launchDirectionDeg.toDouble())
        put("faceAngleDeg", faceAngleDeg.toDouble())
        put("pathAngleDeg", pathAngleDeg.toDouble())
        put("impactOffsetMm", impactOffsetMm.toDouble())
        put("confidence", confidence.toDouble())
        put("validityFlags", validityFlags)
    }.toString()

    companion object {
        const val CURRENT_SCHEMA_VERSION = 1

        const val FACE_ANGLE_VALID = 1 shl 0
        const val PATH_ANGLE_VALID = 1 shl 1
        const val IMPACT_OFFSET_VALID = 1 shl 2
        const val CONFIDENCE_VALID = 1 shl 3
    }
}

File: unity/Assets/PuttVision/Scripts/Telemetry/PuttTelemetry.cs
using System;

namespace PuttVision.Telemetry
{
    [Serializable]
    public sealed class PuttTelemetry
    {
        public const int CurrentSchemaVersion = 1;

        public const int FaceAngleValid = 1 << 0;
        public const int PathAngleValid = 1 << 1;
        public const int ImpactOffsetValid = 1 << 2;
        public const int ConfidenceValid = 1 << 3;

        public int schemaVersion = CurrentSchemaVersion;
        public string shotId = string.Empty;
        public long timestampNs;
        public float ballSpeedMps;
        public float launchDirectionDeg;
        public float faceAngleDeg;
        public float pathAngleDeg;
        public float impactOffsetMm;
        public float confidence;
        public int validityFlags;

        // Native shot-static context. Unity renders these values; it does not reinterpret physics.
        public float startXM;
        public float startYM;
        public float holeDistanceM = 5f;
        public float stimpMeters = 2.8f;
        public float sideSlopePct;
        public float longSlopePct;
        public int terrainProfileId = -1;
        public bool flagstickIn;
        public float grainDirectionDeg;
        public float grainStrength01;
        public float moisture01 = 0.5f;
        public float firmness01 = 0.5f;
        public float trueness01 = 1f;

        public bool HasFaceAngle => (validityFlags & FaceAngleValid) != 0;
        public bool HasPathAngle => (validityFlags & PathAngleValid) != 0;
        public bool HasImpactOffset => (validityFlags & ImpactOffsetValid) != 0;
        public bool HasConfidence => (validityFlags & ConfidenceValid) != 0;

        // Used only by the editor/local smoke simulation. Product motion follows PuttPhysicsFrame.
        public bool IsSimulationReady =>
            schemaVersion == CurrentSchemaVersion &&
            !float.IsNaN(ballSpeedMps) &&
            !float.IsInfinity(ballSpeedMps) &&
            ballSpeedMps >= 0f &&
            !float.IsNaN(launchDirectionDeg) &&
            !float.IsInfinity(launchDirectionDeg);
    }
}
